File: SDRUtils/dashboard/src/features/usd-swaps-tape-v2/utils/ptsScale.ts
// Scale-aware bp rendering for the Package Transaction Spread (PTS) and the
// derived rate spread. SDR PTS values arrive in inconsistent units (decimal
// fraction, percent, or bps), so we anchor the scale to the legs' fixed-rate
// spread: a curve/fly PTS that ties out to the derived spread at a clean scale
// factor is shown in bp at that factor; a single spreadover (no counter-leg)
// falls back to a magnitude band. Mirrors SDRUtils/core/pts_scale.py.
import type { UsdSwapTapeRow } from '../types'
import { EMPTY_VALUE } from '../constants'

const PTS_SCALE_FACTORS = [1, 10, 0.1, 100, 0.01, 1000, 0.001, 10_000, 0.0001]
const SCALE_REL_TOL = 0.05
const DEFAULT_BP_TOL = 0.5

function isFiniteNum(v: unknown): v is number {
  return typeof v === 'number' && Number.isFinite(v)
}

export function findScaleMatch(
  a: number,
  b: number,
  tol = DEFAULT_BP_TOL,
): { factor: number; residual: number } | null {
  if (!Number.isFinite(a) || !Number.isFinite(b)) return null
  if (b === 0) return Math.abs(a) <= tol ? { factor: 1, residual: Math.abs(a) } : null
  let best: { factor: number; residual: number } | null = null
  for (const f of PTS_SCALE_FACTORS) {
    const scaled = b * f
    const residual = Math.abs(a - scaled)
    const fits =
      f === 1
        ? residual <= tol
        : (() => {
            const denom = Math.max(Math.abs(a), Math.abs(scaled))
            return denom < 1e-12 ? residual <= tol : residual / denom <= SCALE_REL_TOL
          })()
    if (fits && (best === null || residual < best.residual)) best = { factor: f, residual }
  }
  return best
}

/** Convert a fixed-rate spread to bp, auto-detecting decimal vs percent rates. */
export function fixedRateSpreadToBp(spread: number, rates: Array<number | null | undefined>): number {
  const finite = rates.filter(isFiniteNum) as number[]
  if (finite.length === 0) return spread * 100
  return spread * (Math.max(...finite.map((r) => Math.abs(r))) < 1 ? 10_000 : 100)
}

function structureKind(row: UsdSwapTapeRow): string {
  return String(row.package_type ?? row.trade_type ?? '').toUpperCase()
}

function isSpreadStructure(kind: string): boolean {
  return (
    kind === 'CURVE' ||
    kind === 'FLY' ||
    kind.endsWith('_CURVE') ||
    kind.endsWith('_FLY')
  )
}

/** Derived rate spread in bp for a curve/fly, given the already-computed
 *  summary rate spread (back-front for curves, 2·belly-wings for flies) and the
 *  legs' fixed rates. Null for non-spread structures or missing data. */
export function derivedSpreadBp(
  row: UsdSwapTapeRow,
  summaryRate: number | null,
): number | null {
  if (!isSpreadStructure(structureKind(row)) || !isFiniteNum(summaryRate)) return null
  const rates = (row.legs_json ?? []).map((l) => l?.fixed_rate ?? null)
  return fixedRateSpreadToBp(summaryRate, rates)
}

/** Reported PTS expressed in bp. When a derived spread is available (curve/fly)
 *  the scale is chosen to tie out to it; otherwise a magnitude band is used
 *  (|pts|<=0.01 -> decimal ×10000, 0.10<=|pts|<=1.0 -> percent ×100). */
export function ptsInBp(pts: number | null | undefined, derivedBp: number | null): number | null {
  if (!isFiniteNum(pts)) return null
  if (pts === 0) return 0
  if (derivedBp !== null && Number.isFinite(derivedBp)) {
    const m = findScaleMatch(Math.abs(derivedBp), Math.abs(pts))
    if (m) return pts * m.factor
  }
  const a = Math.abs(pts)
  if (a <= 0.01) return pts * 10_000
  if (a >= 0.1 && a <= 1.0) return pts * 100
  if (a > 1.0 && a <= 100) return pts // already bp-scale
  return null
}

export function formatBp(n: number | null): string {
  if (n === null || !Number.isFinite(n)) return EMPTY_VALUE
  // Show up to 3 decimals (4 for sub-bp values) so precise ties read as
  // "12.755bps", then trim trailing zeros so round values read as "33.3bps".
  const a = Math.abs(n)
  const digits = a !== 0 && a < 1 ? 4 : 3
  let s = n.toFixed(digits)
  if (s.includes('.')) s = s.replace(/0+$/, '').replace(/\.$/, '')
  return `${s}bps`
}

File: SDRUtils/core/pts_scale.py
# ABOUTME: Canonical scale-aware tie-out between a reported Package
# Transaction Spread (PTS) and a rate-derived spread. SDR PTS values arrive
# in inconsistent units (decimal fraction, percent, or bps — Part 43/45
# notation codes 3/4 plus real-world mis-scaling), so a curve/fly whose PTS
# ties out to its legs' fixed-rate spread at *some* clean scale factor is
# high-confidence evidence of a genuine package.
#
# This mirrors the scale logic in:
#   SDRUtils/analytics/package_confidence.py  (_find_scale_match / _pts_match_signal)
#   SDRUtils/dashboard/.../utils/packageConfidence.ts  (findScaleMatch / ptsMatchSignal)
# Kept in core/ so the *detectors* (packages/curve.py, packages/ptp_grouper.py,
# packages/fly.py) can share ONE tie-out definition with the display overlay
# without an analytics<-packages circular import.
from __future__ import annotations

import math
from typing import Optional, Sequence

# Scale factors for unit-mismatch detection (decimal / percent / bps).
# Order matters: 1 is checked first so an exact match wins on ties.
PTS_SCALE_FACTORS: tuple[float, ...] = (
    1, 10, 0.1, 100, 0.01, 1000, 0.001, 10_000, 0.0001,
)

# Relative tolerance for non-1x scale matches (5%).
SCALE_REL_TOL: float = 0.05

# +/- bp window for a direct (1x) derived-spread vs reported-PTS match.
DEFAULT_PTS_MATCH_BP: float = 0.5


def _finite(v: object) -> bool:
    if v is None:
        return False
    try:
        return math.isfinite(float(v))
    except (TypeError, ValueError):
        return False


def find_scale_match(
    a: float,
    b: float,
    tol: float = DEFAULT_PTS_MATCH_BP,
    scales: Sequence[float] = PTS_SCALE_FACTORS,
) -> Optional[dict]:
    """Try to express *a* as ``b * factor`` for any factor in *scales*.

    Returns ``{"factor": f, "residual": r}`` for the smallest-residual match,
    or ``None``. factor==1 uses the absolute ``tol``; non-1 factors use the
    relative ``SCALE_REL_TOL`` (order-of-magnitude unit mismatch).
    """
    if not math.isfinite(a) or not math.isfinite(b):
        return None
    if b == 0:
        return {"factor": 1.0, "residual": abs(a)} if abs(a) <= tol else None
    best: Optional[dict] = None
    for f in scales:
        scaled = b * f
        residual = abs(a - scaled)
        if f == 1:
            fits = residual <= tol
        else:
            denom = max(abs(a), abs(scaled))
            fits = residual <= tol if denom < 1e-12 else (residual / denom) <= SCALE_REL_TOL
        if fits and (best is None or residual < best["residual"]):
            best = {"factor": float(f), "residual": residual}
    return best


def fixed_rate_spread_to_bp(spread: float, rates: Sequence[Optional[float]]) -> float:
    """Convert a derived fixed-rate spread to bp, auto-detecting whether the
    legs' rates are decimal (0.0384) or percent (3.84)."""
    finite = [float(r) for r in rates if _finite(r)]
    if not finite:
        return spread * 100
    return spread * (10_000 if max(abs(r) for r in finite) < 1 else 100)


def pts_ties_to_spread(
    rate_a: object,
    rate_b: object,
    reported_pts: object,
    bp_tol: float = DEFAULT_PTS_MATCH_BP,
) -> bool:
    """True when the reported PTS ties out to the two legs' fixed-rate spread
    at some clean scale factor (magnitude-based, sign-convention agnostic).

    Returns False whenever any input is missing/non-finite — an *absent* tie
    is never affirmative curve evidence (callers fall back to shape gates).
    """
    if not (_finite(rate_a) and _finite(rate_b) and _finite(reported_pts)):
        return False
    derived_bp = fixed_rate_spread_to_bp(float(rate_b) - float(rate_a),
                                         [float(rate_a), float(rate_b)])
    a = abs(derived_bp)
    b = abs(float(reported_pts))
    if a == 0 and b == 0:
        return True
    if abs(a - b) <= bp_tol:  # direct (1x) match
        return True
    scale = find_scale_match(a, b, bp_tol)
    return scale is not None and scale["factor"] != 1


def pts_ties_to_value(
    derived_bp: object,
    reported_pts: object,
    bp_tol: float = DEFAULT_PTS_MATCH_BP,
) -> bool:
    """Like :func:`pts_ties_to_spread` but the caller supplies the already
    rate-derived spread in bp (e.g. a fly's 2*belly - wings)."""
    if not (_finite(derived_bp) and _finite(reported_pts)):
        return False
    a = abs(float(derived_bp))
    b = abs(float(reported_pts))
    if a == 0 and b == 0:
        return True
    if abs(a - b) <= bp_tol:
        return True
    scale = find_scale_match(a, b, bp_tol)
    return scale is not None and scale["factor"] != 1


def spread_to_bp(value: object) -> Optional[float]:
    """Normalize a single reported spreadover value to basis points via a
    magnitude band (no counter-leg to tie against). Mirrors the accept-bands in
    ``detect_spreadovers``: decimal (|v|<=0.01 -> x10000) or percent
    (0.10<=|v|<=1.0 -> x100). The ambiguous middle (0.01, 0.10) and clearly
    erroneous magnitudes (>1.0) return None."""
    if not _finite(value):
        return None
    v = float(value)
    a = abs(v)
    if a == 0:
        return 0.0
    if a <= 0.01:
        return v * 10_000.0
    if 0.10 <= a <= 1.0:
        return v * 100.0
    return None
